### ISSUE TITLE
Unit testing component with debounceTime rxjs operator is way too hard

### ISSUE THREAD

**Issue by jpzwarte** [angular/angular#25457]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

```
        it('should autocomplete schools', fakeAsync(() => {
          spectator.click('eb-select2');
          spectator.typeInElement('ba', 'input.eb-filter');

          // Wait for the debounceTime(500)
          tick(500);
          spectator.fixture.whenStable().then(() => {
            spectator.detectChanges();

            const schools = spectator
              .queryAll('.eb-select-dropdown .vo-name')
              .map(elm => elm.textContent);

            expect(schools).toEqual(['Bar', 'Baz']);
          });
        }));
```

That code (uses the spectator library, but that just wraps the TestBed API) is required to test a component that has an autocomplete feature that uses ```debounceTime(500)```

```
  ngOnInit() {
    this.schools = this.schoolFilter.pipe(
      distinctUntilChanged(),
      debounceTime(500),
      startWith(''),
      switchMap(term => {
       // etc.
      })
    );
```

Testing code that uses the ```debounceTime``` operator should not be this difficult. By difficult i mean you have to:
1. Wrap the test in ```fakeAsync```
2. Advance time using ```tick(500)```
3. Then wait until the fixture is table
4. Finally ```detectChanges```

The docs at https://angular.io/guide/testing#fake-async make no mention of having to use ```whenStable()```. In fact it says the opposite: "The fakeAsync function enables a linear coding style by running the test body in a special fakeAsync test zone. The test body appears to be synchronous. There is no nested syntax (like a Promise.then()) to disrupt the flow of control."

**Comment by jpzwarte** [comment]
Also perhaps improve ```tick()``` to do the other steps that are required automatically?

**Comment by meriturva** [comment]
On my case i have debounce time value on an app `configuration` injected to main module set to 500.

When i run tests i have a `mockConfiguration` class that has just debounce time value to 0. So no more values on many services and a simple way to control it on tests.

**Comment by JiaLiPassion** [comment]
@jpzwarte, in this PR, we have a more clear explanation about how to use `fakeAsync` to test `rxjs timer operator`. https://github.com/angular/angular/pull/23117.
Here is an example.

```ts
 import '../../lib/rxjs/rxjs-fake-async';
    it('should get date diff correctly', (done) => {
      fakeAsyncTestZone.run(() => {
        let result = null;
        const observable = new Observable((subscribe: any) => {
          subscribe.next('hello');
        });
        observable.delay(1000).subscribe(v => {
          result = v;
        });
        expect(result).toBeNull();
        testZoneSpec.tick(1000);
        expect(result).toBe('hello');
        done();
      });
    });
```

**Comment by Abrissirba** [comment]
I don't really understand the your example @JiaLiPassion. Where should I import rxjs-fake-async from?

**Comment by JiaLiPassion** [comment]
@Abrissirba, sorry, it should be

```
import 'zone.js/dist/zone-patch-rxjs-fake-async';
```

**Comment by roldengarm** [comment]
This does not compile @JiaLiPassion , I can't import that

**Comment by fagia** [comment]
@jpzwarte as very well explained here: https://stackoverflow.com/a/41658320/1042989, you shouldn't mix `fakeAsync` with `fixture.whenStable` but, to get your example working, it should be enough to have a `tick` greater than the `debounceTime`(ie: if you're debouncing for 500ms try with `tick(600)`)

**Comment by nasreddineskandrani** [comment]
Hi,
Please find a working solution in this stackblitz:
https://stackblitz.com/edit/angular-component-observable-testing-esguy4?file=app%2Fm1%2Fc1.component.ts

The 2 first tests show it's not working:
1.test debounce - (option 1) do nothing
2.test debounce - (option 2) fakeAsync/flush

The 2 last show **it's working**:
3. test debounce - (option 3)  fakeAsync/tick (the tick myst be bigger than the debounce time)
4 .test debounce - (option 4) rxjs-marbles/flush

I recommend to use 'option 4'

Thank you

**Comment by thw0rted** [comment]
@nasreddineskandrani The original question was mistaken in using `fixture.whenStable`.  It also complained about "difficulty" that has been addressed -- you can call `fixture.detectChanges()` but you can also set `autoDetectChanges`.

The one remaining good point is that you have to know a magic number (the debounce time) to write the test.  I tried to [update your example](https://stackblitz.com/edit/zone-macrotask-flush-example?file=src/app/m1/c1.component.spec.ts) to get rid of the magic number, using `flush` as I suggested above. It doesn't seem to work, though.  I'm not sure why: [the docs](https://angular.io/api/core/testing/flush) say that `flush` "{drains} the macrotask queue until it is empty".  If I read the rxjs source correctly, it looks like `debounceTime` is just a thin wrapper for `AsyncScheduler`/`AsyncTask`, which uses `setInterval` under the hood, which is a macrotask.  If that's the case, why wouldn't my linked example work?  It uses `flushMicrotasks()` so that the (Promise-based) subscription set up in `ngOnInit` can start the pipeline, and I verified that the inner logic of `debounceTime` is set up before `flush()` is called.  Does `flush` ignore intervals?

**Comment by jpzwarte** [comment]
@thw0rted iirc `whenStable` was required because the component uses `ngModel` (which is async by default).

**Comment by thw0rted** [comment]
I'm having trouble finding a reference in the docs, but `whenStable` is only supposed to be called inside `waitForAsync`, not `fakeAsync`.  Basically, each test has to choose (`waitForAsync` + `whenStable`) or (`fakeAsync` + `flush` + `tick`), or of course you can just use Jasmine's native async/promise handling.  If you are testing a component that uses `ngModel` (or in the linked Stackblitz, `(input)` event handling and template-string insertion) using `fakeAsync`, then you just have to call `tick` or `flush` to make sure the component updates, rather than using `whenStable`.

**Comment by nasreddineskandrani** [comment]
@thw0rted is not that bad to share the number with the config in the test :)
I made it work with other approaches based on `TestScheduler` without `FakeAsync` then i remove my initial answer and focused in one easy with `FakeAsync` to dont confuse readers that are looking for a working solution. You can ping me in gitter if you want to talk about `TestSCheduler` from rxjs.

@jpzwarte this is an example working with `ngModel`
https://stackblitz.com/edit/angular-debouncetime-rxjs-testing-ngmaterial

**Comment by nasreddineskandrani** [comment]
@thw0rted 
> The one remaining good point is that you have to know a magic number (the debounce time) to write the test. I tried to update your example to get rid of the magic number, using flush as I suggested above. It doesn't seem to work, though. I'm not sure why: the docs say that flush "{drains} the macrotask queue until it is empty". If I read the rxjs source correctly, it looks like debounceTime is just a thin wrapper for AsyncScheduler/AsyncTask, which uses setInterval under the hood, which is a macrotask. If that's the case, why wouldn't my linked example work? It uses flushMicrotasks() so that the (Promise-based) subscription set up in ngOnInit can start the pipeline, and I verified that the inner logic of debounceTime is set up before flush() is called. Does flush ignore intervals?

its not using `interval` its using `date.now`
look here https://stackoverflow.com/questions/50759469/how-do-i-mock-rxjs-6-timer

**Comment by nasreddineskandrani** [comment]
@thw0rted `setTimeout` and `setInterval` are not microtasks so that's why `flush` is not working. Read here:
https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide

**Comment by nasreddineskandrani** [comment]
the `flush` from `rxjs-marbles` is working (use `flush` from `TestScheduler` that inhrit `VirtualTimeScheduler` from rxjs)
https://stackblitz.com/edit/zone-macrotask-flush-example-pscagb?file=src%2Fapp%2Fm1%2Fc1.component.spec.ts

Why the `flush` from angular is no working? the question remains... for now no answer here

**Comment by thw0rted** [comment]
Apologies in advance, this wound up being kind of stream-of-consciousness:

Your Jake Archibald link references [the formal definition of setTimeout](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#timer-initialisation-steps).  In short, it records a bunch of stuff, creates a task (step 7), records more stuff, "yields" the task ID (step 14), then continues "running in parallel", "waits" for a time (step 15), then *actually puts the task on the global queue* in the last step.

So, per the spec, when you call a method that *eventually* puts a task on the queue, the task is "prepared" before the method returns execution context to the caller, but the method "forks" and continues running, "sleeping" until the specified time has elapsed, *then* actually putting the task on the queue.  (The only difference between `setTimeout` and `setInterval` here is that the latter creates a task that calls `setInterval` again after the task function is invoked, per step 7.3)

This is where the confusion happens.  `flush` says it "drains the queue" of macrotasks.  But there are no macrotasks *on the queue* when I call `flush()`, there are background "threads" sleeping, *waiting* to put a task on the queue -- or at least, there would be, if we weren't in the `fakeAsync` zone.  I think "the queue" referenced in the Angular docs is not the global macrotask queue, but rather a `fakeAsync`-specific cheat-sheet of pending timers, that *will* publish tasks to the queue some (simulated!) time later.

Now, the question becomes, if this "cheat sheet" of pending timers (that aren't "queued tasks" yet, but will be) has both timeouts and intervals, what does `flush` actually do?  It seems to run all timeouts, but none of the intervals -- actually, I just found https://github.com/angular/angular/issues/34742#issuecomment-597259879 which confirms it:

> currently the flush() API only flush non-periodic macroTasks such as setTimeout(), so for now, please use tick() to make it work, and I am currently working on adding new APIs to make this work easily

----

TL;DR: `flush` is confirmed to only run "non-periodic macrotasks".  The docs should be updated to describe the exact current behavior.  This ticket can probably be closed in favor of #34742 (which promises "new APIs" to manage periodic microtasks) because the original complaint -- making it easier to test `debounceTime` -- is addressed by a combination of `autoDetectChanges` and the hypothetical periodic microtask "flush-like" API.

**Comment by nasreddineskandrani** [comment]
@thw0rted amazing explanation.
 I know we are a little far from the initial problem but seriously i think it's legit to say we made work with `tick` we wanted to know why the solution `flush` is not working. 
Again amazing work. respect!

**Comment by AlaaEl-DinAhmed** [comment]
> On my case i have debounce time value on an app `configuration` injected to main module set to 500.
> 
> When i run tests i have a `mockConfiguration` class that has just debounce time value to 0. So no more values on many services and a simple way to control it on tests.

How do you implement it please?

**Comment by nasreddineskandrani** [comment]
@AlaaEl-DinAhmed its really far from me this topic. What i can assure you is that you can test wverything you need without mocking config by reading my article here:
https://medium.com/front-end-tricks/testing-with-fakeasync-in-angular-d7d0550994a4

If you don't find an answer in my article. What you can is create an example in stackblitz with the problem you are facing . I'll solve it.

**Comment by mgechev** [comment]
This seems like an inconvenient use case, but unfortunately we can't capture all the use cases of Angular with third-party libraries (such as RxJS) in the documentation. I'll close the issue for now and leave it as an opportunity for a community article.

**Comment by thw0rted** [comment]
@mgechev Rx is hardly a "third party" library, and is used extensively within Angular.  The [issue I suggested as a replacement](https://github.com/angular/angular/issues/34742) has now also been closed without being fixed, and [the change I suggested to the docs](https://github.com/angular/angular/issues/25457#issuecomment-728876345) has not been made -- [the page](https://angular.io/api/core/testing/flush) still just says "draining the macrotask queue" and doesn't mention the distinction between periodic vs non-periodic macrotasks.

I think a "community article" could help but I also think that if Angular wants its users to create unit tests for components then this needs to *be easy*.  Debouncing input using RxJS `debounce` is like second-day-of-Angular-school stuff, everybody probably does it at least once in a project of reasonable size, and testing it will continue to suck until periodic-macrotask management improves.  Worse, last time I checked, the Angular Material / Material Components test harness library can't be used with debounce/interval due to this issue (see my reference above).

**Comment by GerkinDev** [comment]
My 2 cents on this: Assuming you're using jest, add the following at the very top of your test to replace the `asyncScheduler` with `queueScheduler` during test, properly drained by `tick`:

```
jest.mock( 'rxjs', () => ( {
	...jest.requireActual( 'rxjs' ),
	asyncScheduler: jest.requireActual( 'rxjs' ).queueScheduler,
} ) );
```
